File: frontend/src/lib/blocks.ts
// Shared Block schema (mirrors backend/schemas/blocks.py).
// An assistant response is an ordered array of Blocks plus a results map keyed
// by resultId. The frontend renders blocks in order and resolves resultId -> data.

export const BLOCK_TYPES = [
  "markdown",
  "metric",
  "chart",
  "histogram",
  "table",
  "recommendation",
  "customer_card",
  "product_card",
  "order_card",
] as const;

export type BlockType = (typeof BLOCK_TYPES)[number];

export type ChartType = "line" | "bar" | "scatter";

export interface SqlResult {
  resultId: string;
  columns: string[];
  rows: unknown[][];
  n_rows: number;
}

// ---- blocks ----
export interface MarkdownBlock {
  id: string;
  type: "markdown";
  content: string;
}

export interface MetricBlock {
  id: string;
  type: "metric";
  resultId?: string;
  label: string;
  valueKey?: string;
  rowIndex?: number;
  change?: string;
  trend?: "up" | "down" | "neutral";
}

export interface ChartBlock {
  id: string;
  type: "chart";
  resultId: string;
  chartType: ChartType;
  xKey?: string;
  series?: { dataKey: string; label?: string }[];
  title?: string;
}

export interface HistogramBlock {
  id: string;
  type: "histogram";
  resultId: string;
  dataKey: string;
  bins?: number;
  title?: string;
}

export interface TableBlock {
  id: string;
  type: "table";
  resultId?: string;
  columns?: string[];
  rows?: unknown[][];
  title?: string;
}

export interface RecommendationBlock {
  id: string;
  type: "recommendation";
  resultId?: string;
  title?: string;
  text: string;
  reason?: string;
}

export interface CustomerCardBlock {
  id: string;
  type: "customer_card";
  resultId?: string;
  customerId: string;
}

export interface ProductCardBlock {
  id: string;
  type: "product_card";
  resultId?: string;
  productId: string;
}

export interface OrderCardBlock {
  id: string;
  type: "order_card";
  resultId?: string;
  orderId: string;
}

export type Block =
  | MarkdownBlock
  | MetricBlock
  | ChartBlock
  | HistogramBlock
  | TableBlock
  | RecommendationBlock
  | CustomerCardBlock
  | ProductCardBlock
  | OrderCardBlock;

export interface AssistantResponse {
  blocks: Block[];
  results: Record<string, SqlResult>;
}

// ---------------------------------------------------------------------------
// Validation (never throws; malformed blocks are dropped)
// ---------------------------------------------------------------------------
function isObj(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

export function validateBlocks(raw: unknown): Block[] {
  if (!Array.isArray(raw)) return [];
  const out: Block[] = [];
  let i = 0;
  for (const item of raw) {
    if (!isObj(item)) continue;
    const type = item["type"];
    if (typeof type !== "string" || !(BLOCK_TYPES as readonly string[]).includes(type)) continue;
    const id = typeof item["id"] === "string" ? item["id"] : `b${i}`;
    const block = { ...item, id } as Block;
    if (!isValidBlock(block)) continue;
    out.push(block);
    i++;
  }
  return out;
}

function isValidBlock(b: Block): boolean {
  switch (b.type) {
    case "markdown":
      return typeof b.content === "string";
    case "metric":
      return typeof b.label === "string";
    case "chart":
      return (
        typeof b.resultId === "string" &&
        (b.chartType === "line" || b.chartType === "bar" || b.chartType === "scatter")
      );
    case "histogram":
      return typeof b.resultId === "string" && typeof b.dataKey === "string";
    case "table":
      return true; // table can resolve from resultId OR inline columns/rows
    case "recommendation":
      return typeof b.text === "string";
    case "customer_card":
    case "product_card":
    case "order_card":
      return typeof (b as { customerId?: string; productId?: string; orderId?: string }).customerId === "string" ||
             typeof (b as { customerId?: string; productId?: string; orderId?: string }).productId === "string" ||
             typeof (b as { customerId?: string; productId?: string; orderId?: string }).orderId === "string";
    default:
      return false;
  }
}

// ---------------------------------------------------------------------------
// Result resolution
// ---------------------------------------------------------------------------
export function resolveResult(
  results: Record<string, SqlResult>,
  resultId?: string
): SqlResult | null {
  if (!resultId) return null;
  return results[resultId] ?? null;
}

/** Resolve a metric's single value from its result. */
export function resolveMetricValue(block: MetricBlock, results: Record<string, SqlResult>): string | null {
  const res = resolveResult(results, block.resultId);
  if (!res || !res.rows.length) return null;
  const row = res.rows[block.rowIndex ?? 0];
  if (block.valueKey) {
    const idx = res.columns.indexOf(block.valueKey);
    if (idx >= 0 && row[idx] != null) return String(row[idx]);
  }
  // fallback: first non-null cell in the first row
  const first = row.find((v) => v != null);
  return first != null ? String(first) : null;
}

/** Resolve a table block's columns/rows (result first, then inline). */
export function resolveTable(
  block: TableBlock,
  results: Record<string, SqlResult>
): { columns: string[]; rows: unknown[][]; title?: string } | null {
  if (block.resultId) {
    const res = resolveResult(results, block.resultId);
    if (res) return { columns: res.columns, rows: res.rows, title: block.title };
  }
  if (block.columns && block.rows) {
    return { columns: block.columns, rows: block.rows, title: block.title };
  }
  return null;
}

/** Resolve chart data rows from a result. */
export function resolveChartData(
  block: ChartBlock,
  results: Record<string, SqlResult>
): SqlResult | null {
  return resolveResult(results, block.resultId);
}

File: backend/schemas/blocks.py
"""Shared, validated Block schema for copilot responses.

This is the strict UI contract between the LLM/backend and the frontend.
An assistant response is an ordered array of ``blocks`` plus a ``results`` map
of MCP query results referenced by ``resultId``.

The frontend mirrors these types in TypeScript (see frontend/src/lib/blocks.ts).
"""
from __future__ import annotations

from typing import Any, Literal, Union

from pydantic import BaseModel, Field, ValidationError, field_validator, model_validator

# ---------------------------------------------------------------------------
# Structured result (as returned by the MCP run_sql tool)
# ---------------------------------------------------------------------------
class SqlResult(BaseModel):
    resultId: str
    columns: list[str]
    rows: list[list[Any]]
    n_rows: int = Field(default=0)

    @model_validator(mode="after")
    def _sync_count(self) -> "SqlResult":
        if not self.n_rows:
            self.n_rows = len(self.rows)
        return self


# ---------------------------------------------------------------------------
# Blocks
# ---------------------------------------------------------------------------
class BaseBlock(BaseModel):
    id: str = Field(..., min_length=1)
    type: str


class MarkdownBlock(BaseBlock):
    type: Literal["markdown"]
    content: str


class MetricBlock(BaseBlock):
    type: Literal["metric"]
    resultId: str | None = None
    label: str
    valueKey: str | None = None  # column to read the value from
    rowIndex: int = 0
    change: str | None = None
    trend: Literal["up", "down", "neutral"] | None = None


class ChartBlock(BaseBlock):
    type: Literal["chart"]
    resultId: str
    chartType: Literal["line", "bar", "scatter"]
    xKey: str | None = None
    series: list[dict[str, str]] = Field(default_factory=list)  # [{"dataKey","label"}]
    title: str | None = None


class HistogramBlock(BaseBlock):
    type: Literal["histogram"]
    resultId: str
    dataKey: str
    bins: int = 10
    title: str | None = None


class TableBlock(BaseBlock):
    type: Literal["table"]
    resultId: str | None = None
    columns: list[str] | None = None
    rows: list[list[Any]] | None = None
    title: str | None = None


class RecommendationBlock(BaseBlock):
    type: Literal["recommendation"]
    resultId: str | None = None
    title: str | None = None
    text: str
    reason: str | None = None


class CustomerCardBlock(BaseBlock):
    type: Literal["customer_card"]
    resultId: str | None = None
    customerId: str


class ProductCardBlock(BaseBlock):
    type: Literal["product_card"]
    resultId: str | None = None
    productId: str


class OrderCardBlock(BaseBlock):
    type: Literal["order_card"]
    resultId: str | None = None
    orderId: str


Block = Union[
    MarkdownBlock,
    MetricBlock,
    ChartBlock,
    HistogramBlock,
    TableBlock,
    RecommendationBlock,
    CustomerCardBlock,
    ProductCardBlock,
    OrderCardBlock,
]

BLOCK_TYPES = {
    "markdown", "metric", "chart", "histogram", "table",
    "recommendation", "customer_card", "product_card", "order_card",
}


class ChatResponse(BaseModel):
    """The full assistant reply: ordered blocks + a results map keyed by resultId."""

    blocks: list[Block] = Field(default_factory=list)
    results: dict[str, SqlResult] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
def validate_blocks(raw: Any) -> list[Block]:
    """Validate raw LLM-produced blocks, dropping malformed/unknown ones.

    Never raises; malformed entries are replaced with a safe markdown fallback
    so an unknown block cannot crash rendering.
    """
    if not isinstance(raw, list):
        return []
    out: list[Block] = []
    for i, item in enumerate(raw):
        if not isinstance(item, dict):
            continue
        btype = item.get("type")
        if btype not in BLOCK_TYPES:
            continue
        # guaranteed id
        item = dict(item)
        item.setdefault("id", f"b{i}")
        try:
            block = _construct(btype, item)
            out.append(block)
        except (ValidationError, TypeError, ValueError):
            continue
    return out


def _construct(btype: str, item: dict[str, Any]) -> Block:
    model: type[BaseModel] = {
        "markdown": MarkdownBlock,
        "metric": MetricBlock,
        "chart": ChartBlock,
        "histogram": HistogramBlock,
        "table": TableBlock,
        "recommendation": RecommendationBlock,
        "customer_card": CustomerCardBlock,
        "product_card": ProductCardBlock,
        "order_card": OrderCardBlock,
    }[btype]
    return model(**item)


def blocks_from_dicts(raw: list[dict[str, Any]]) -> ChatResponse:
    """Build a ChatResponse from raw blocks + results dicts (loose input)."""
    return ChatResponse(blocks=validate_blocks(raw), results={})
